bindings/cpu_profiler.cc

```diff
@@ -52,6 +57,171 @@ v8::CpuProfilingLoggingMode GetLoggingMode() {
+  self->listeners.emplace(std::string(profile_id), cb);
+
+  if (self->listeners.size() == 1) {
+    uv_timer_set_repeat(&self->timer, self->period_ms);
+    uv_timer_start(&self->timer, callback, 0, self->period_ms);
+  }
+}
+
+void HeapStatisticsTimer::remove_listener(uv_timer_t* handle, const char* profile_id, std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)>& cb) {
+  auto self = static_cast<HeapStatisticsTimer*>(handle->data);
+  self->listeners.erase(std::string(profile_id));
+
+  if (self->listeners.size() == 0) {
+    uv_timer_stop(&self->timer);
+  }
+};
+
+class Profiler {
+public:
+  std::unordered_map<std::string, SentryProfile*> active_profiles = std::unordered_map<std::string, SentryProfile*>();
+
+  v8::CpuProfiler* cpu_profiler;
+  HeapStatisticsTimer heap_statistics_timer = HeapStatisticsTimer(uv_default_loop());
+
+  explicit Profiler(const napi_env& env, v8::Isolate* isolate) :
+    cpu_profiler(
+      v8::CpuProfiler::New(isolate, kNamingMode, GetLoggingMode())) {
+    napi_add_env_cleanup_hook(env, DeleteInstance, this);
+  }
+
+  static void DeleteInstance(void* data);
+};
+
+void Profiler::DeleteInstance(void* data) {
+  Profiler* profiler = static_cast<Profiler*>(data);
+  profiler->cpu_profiler->Dispose();
+  delete profiler;
+}
+
+class SentryProfile {
+private:
+
+  uint64_t started_at = 0;
+  uint16_t heap_measurement_index = 0;
+  std::vector<uint64_t> heap_stats_ts = std::vector<uint64_t>(300);
+  std::vector<uint64_t> heap_stats_usage = std::vector<uint64_t>(300);
+  std::function<void(uv_timer_t*, uint64_t, v8::HeapStatistics&)> memory_sampler_cb = [this](uv_timer_t* timer, uint64_t ts, v8::HeapStatistics& stats) {
+    heap_stats_ts.insert(heap_stats_ts.begin() + heap_measurement_index, ts - started_at);
+    heap_stats_usage.insert(heap_stats_usage.begin() + heap_measurement_index, static_cast<uint64_t>(stats.used_heap_size()));
+    ++heap_measurement_index;
+    };
+
+public:
+  const char* id;
+  explicit SentryProfile(const char* id) : started_at(uv_hrtime()), heap_measurement_index(0), id(id) {}
+  ProfileStatus status = ProfileStatus::kNotStarted;
+
+  std::vector<uint64_t>& heap_usage_timestamps();
+  std::vector<uint64_t>& heap_usage_values();
+  uint16_t heap_usage_entries_count();
+
+  void Start(Profiler* profiler);
+  v8::CpuProfile* Stop(Profiler* profiler);
+};
+
+void SentryProfile::Start(Profiler* profiler) {
+  v8::Local<v8::String> profile_title = v8::String::NewFromUtf8(v8::Isolate::GetCurrent(), id, v8::NewStringType::kNormal).ToLocalChecked();
+
+  started_at = uv_hrtime();
+
+  // Initialize the CPU Profiler
+  profiler->cpu_profiler->StartProfiling(profile_title, {
+      v8::CpuProfilingMode::kCallerLineNumbers,
+      v8::CpuProfilingOptions::kNoSampleLimit,
+      kSamplingInterval
+    });
+
+
+  // listen for memory sample ticks
+  profiler->heap_statistics_timer.add_listener(&profiler->heap_statistics_timer.timer, id, memory_sampler_cb);
```

Comment:
`add_listener` is declared as a static method but you're calling it on an instance of `HeapStatisticsTimer` here. If you made it an instance method instead of a static method you wouldn't have to explicitly pass in the `timer`, since that's a member variable of the instance.